contracts/modules/v2/StakingPool.sol

```diff
@@ -173,145 +174,238 @@ contract StakingPool is IStakingPool, ERC721 {
     // SLOAD
     uint _activeStake = activeStake;
     uint _rewardPerSecond = rewardPerSecond;
-    uint _accRewardPerShare = accRewardPerPoolShare;
     uint _stakeSharesSupply = stakeSharesSupply;
     uint _rewardsSharesSupply = rewardsSharesSupply;
-    uint _lastRewardUpdate = lastRewardUpdate;
-    uint _firstActiveGroupId = firstActiveGroupId;
-
-    // first group for the current timestamp
-    uint targetGroupId = block.timestamp / GROUP_SIZE;
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+    uint _lastAccNxmUpdate = lastAccNxmUpdate;
 
     while (_firstActiveBucketId < currentBucketId) {
 
       ++_firstActiveBucketId;
       uint bucketEndTime = _firstActiveBucketId * BUCKET_SIZE;
-      uint elapsed = bucketEndTime - _lastRewardUpdate;
+      uint elapsed = bucketEndTime - _lastAccNxmUpdate;
 
-      _accRewardPerShare += elapsed * _rewardPerSecond / _rewardsSharesSupply;
-      _lastRewardUpdate = bucketEndTime;
+      // todo: should be allowed to overflow?
+      // todo: handle division by zero
+      _accNxmPerRewardsShare += elapsed * _rewardPerSecond / _rewardsSharesSupply;
+      _lastAccNxmUpdate = bucketEndTime;
+      // TODO: use _firstActiveBucketId before incrementing it?
       _rewardPerSecond -= poolBuckets[_firstActiveBucketId].rewardPerSecondCut;
 
       // should we expire a group?
       if (
         bucketEndTime % GROUP_SIZE != 0 ||
-        _firstActiveGroupId == targetGroupId
+        _firstActiveGroupId == currentGroupId
       ) {
         continue;
       }
 
+      // todo: handle _firstActiveGroupId = 0 case
+
       // SLOAD
-      StakeGroup memory group = stakeGroups[_firstActiveGroupId];
+      Group memory group = groups[_firstActiveGroupId];
```

Comment:
This code is located in the `updateGroups` function and the contents of the `_firstActiveGroupId` is stale at the moment. The group here is actually being expired, so calling it `firstActiveGroup` is actually misleading. I'm renaming it to `expiringGroup`.

Reply:
`termToEnd` is even better since all active terms are expiring/ending.
`ExpiredGroup` -> `EndedTerm`

Having a term end (reaching maturity date) does not mean it is expired (which signals a finality). The end just means that the corresponding stake shares cannot be used to sell further covers. The users can still choose to extend their deposit until one of the next terms.